README.md
# elixir-pdf-generator

A wrapper for both wkhtmltopdf and chrome-headless plus PDFTK (adds in
encryption) for use in Elixir projects.

# Latest release v0.6.0 on 2019-12-17

- 0.6.0
  - introducting `make` as build tool (optional) for chromium binaries
    (puppeteer)
  - **BUGFIX:** documentation: option `pagesize` requires string argument
    (for example `"letter"` or `"A4"`)
  - updated some npm dependencies for chromium

For a proper changelog, see [CHANGES](CHANGES.md)

# Usage

_Hint:_ In IEX, `h PdfGenerator.generate` is your friend.

Add this to your dependencies in your mix.exs:

```Elixir
    def application do
        [applications: [
            :logger,
            :pdf_generator # <-- add this for Elixir <= 1.4
        ]]
    end

    defp deps do
        [
            # ... whatever else
            { :pdf_generator, ">=0.6.0" }, # <-- and this
        ]
    end
```

If you want to use a locally-installed chromium in **RELEASES** (think `mix
release`), alter your mixfile to let `make` take care of compilation and
dependency-fetching:

```Elixir
defp deps do
  [
    # ... whatever else
    { :pdf_generator, ">=0.6.0", compile: "make chrome" }
  ]
end
```

This will embed a **300 MB** (yes, that large) Chromium binary into your priv folder
which will survive packaging as Erlang release. This _can_ be handy as this will
run on slim Alpine docker images with just NodeJS installed.

The recommended way still is to install Chromium/Puppeteer globally and set the
`prefer_system_executable: true` option when generating PDFs.

In development: While this usually works, it unfortunately leads to
pdf_generator to be compiled all the time again and again due to my bad Makefile
skills. Help is very much appreciated.

# Try it out

Pass some HTML to PdfGenerator.generate:

```Elixir
$ iex -S mix

html = "<html><body><p>Hi there!</p></body></html>"
# be aware, this may take a while...
{:ok, filename}    = PdfGenerator.generate(html, page_size: "A5")
{:ok, pdf_content} = File.read(filename)

# or, if you prefer methods that raise on error:
filename = PdfGenerator.generate!(html, generator: :chrome)
```

Or, pass some URL

```Elixir
PdfGenerator.generate {:url, "http://google.com"}, page_size: "A5"
```

Or use the bang-methods:

```Elixir
filename   = PdfGenerator.generate! "<html>..."
pdf_binary = PdfGenerator.generate_binary! "<html>..."
```

## Chrome

Or, use **chrome-headless**.

Unless your mixfile sais `{:pdf_generator, ">=6.0.0", compile: "make chrome"}`
Chrome won't be installed into your application. Please set the
`prefer_system_executable: true` option in this case.

```Elixir
html_works_too = "<html><body><h1>Minimalism!"
{:ok, filename} = PdfGenerator.generate html_works_too, generator: :chrome, prefer_system_executable: true
```

## Docker

If using chrome in a superuser/root environment (read: **docker**), make sure to
pass an option to chrome to disable sandboxing. And be aware of the implications.

```Elixir
html_works_too = "<html><body><h1>I need Docker, baby docker is what I need!"
{:ok, filename} = PdfGenerator.generate html_works_too, generator: :chrome, no_sandbox: true, page_size: "letter"
```

# System prerequisites 

It's either 

* wkhtmltopdf or 

* nodejs (for Chrome-headless/Puppeteer)

## chrome-headless

This will allow you to make more use of Javascript and advanced CSS as it's just
your Chrome/Chromium browser rendering your web page as HTML and printing it as
PDF. Rendering _tend_ to be a bit faster than with wkhtmltopdf. The price tag is
that PDFs printed with chrome/chromium are usually considerably bigger than
those generated with wkhtmltopdf.

### global install (great for Docker images)

Run `npm -g install chrome-headless-render-pdf puppeteer`. 

This requires [nodejs](https://nodejs.org), of course. This will install a
recent chromium and chromedriver to run Chrome in headless mode and use this
browser and its API to print PDFs globally on your machine.
   
If you prefer a project-local install, use the `compile: "make chrome"` option
in your mixfile's dependency-line.

On some machines, this doesn't install Chromium and fails. Here's how to get
this running on Ubuntu 18:
   
```bash
DEBIAN_FRONTEND=noninteractive PUPPETEER_SKIP_CHROMIUM_DOWNLOAD=TRUE \
  apt-get install -y chromium-chromedriver \
  && npm -g install chrome-headless-render-pdf puppeteer
```

### local install

Run `make priv/node_modules`. This requires both `nodejs` (insallation see
above) and `make`. 

Or, run `cd priv && npm install`
   
## wkhtmltopdf

2. Download wkhtmltopdf and place it in your $PATH. Current binaries can be
   found here: http://wkhtmltopdf.org/downloads.html
   
   For the impatient (Ubuntu 18.04 Bionic Beaver):
   
   ```
   apt-get -y install xfonts-base xfonts-75dpi \
    && wget https://downloads.wkhtmltopdf.org/0.12/0.12.5/wkhtmltox_0.12.5-1.bionic_amd64.deb \
    && dpkg -i wkhtmltox_0.12.5-1.bionic_amd64.deb
   ```
   
   For other distributions, refer to http://wkhtmltopdf.org/downloads.html – For
   example, replace `bionic` with `xenial` if you're on Ubuntu 16.04.
   
## optional dependencies

3. _optional:_ Install `xvfb` (shouldn't be required with the binary mentioned above):

   To use other wkhtmltopdf executables comiled with an unpatched Qt on systems
   without an X window server installed, please install `xvfb-run` from your
   repository (on Debian/Ubuntu: `sudo apt-get install xvfb`).
   
   I haven't heard any feedback of people using this feature since a while since
   the wkhtmltopdf projects ships ready-made binaries. I will deprecate this
   starting in `0.6.0` since, well, YAGNI.

4. _optional:_ Install `pdftk` via your package manager or homebrew. The project
   page also contains a Windows installer. On Debian/Ubuntu just type:
   `apt-get -y install pdftk`

# Options and Configuration

This module will automatically try to finde both `wkhtmltopdf` and `pdftk` in
your path. But you may override or explicitly set their paths in your
`config/config.exs`.

```Elixir
config :pdf_generator,
    wkhtml_path:    "/usr/bin/wkhtmltopdf",   # <-- this program actually does the heavy lifting
    pdftk_path:     "/usr/bin/pdftk"          # <-- only needed for PDF encryption
```

or, if you prefer chrome-headless

```
config :pdf_generator,
    use_chrome: true,                           # <-- make sure you installed node/puppeteer
    prefer_system_executable: true              # <-- set this if you installed the NPM dependencies globally
    raise_on_missing_wkhtmltopdf_binary: false, # <-- so the app won't complain about a missing wkhtmltopdf
```

## More options

- `filename` - filename for the output pdf file (without .pdf extension, defaults to a random string)

- `page_size`:
  *  defaults to `"A4"`, see `wkhtmltopdf` for more options
  * `"letter"` (for US letter) be translated to 8x11.5 inches (currently, only in chrome).

- `open_password`:    requires `pdftk`, set password to encrypt PDFs with

- `edit_password`:    requires `pdftk`, set password for edit permissions on PDF

- `shell_params`:     pass custom parameters to `wkhtmltopdf`. **CAUTION: BEWARE OF SHELL INJECTIONS!**

- `command_prefix`:   prefix `wkhtmltopdf` with some command or a command with options
                      (e.g. `xvfb-run -a`, `sudo` ..)

- `delete_temporary`: immediately remove temp files after generation

## Contribution; how to run tests

You're more than welcome ot submit patches. Please run `mix test` to ensure at bit of stability. Tests require a full-fledged environment, with all of `wkhtmltopdf`, `xvfb` and `chrome-headless-render-pdf` available path. Also make to to have run `npm install` in the app's base directory (will install chrome-headless-render-pdf non-globally in there). With all these installed, `mix test` should run smoothly.

_Hint_: Getting `:enoent` errors ususally means that chrome or xvfb couldn't be run. Yes, this should output a nicer error.

## Heroku Setup

If you want to use this project on heroku, you can use buildpacks instead of binaries
to load `pdftk` and `wkhtmltopdf`:
```
https://github.com/fxtentacle/heroku-pdftk-buildpack
https://github.com/dscout/wkhtmltopdf-buildpack
https://github.com/HashNuke/heroku-buildpack-elixir
https://github.com/gjaldon/phoenix-static-buildpack
```

__note:__ The list also includes Elixir and Phoenix buildpacks to show you that they
must be placed after `pdftk` and `wkhtmltopdf`. It won't work if you load the
Elixir and Phoenix buildpacks first.

## Running non-patched wkhtmltopdf headless

This section only applies to `wkhtmltopdf` users using wkhtmltopdf w/o the qt patch. If you are using the latest 0.12 binaries from https://downloads.wkhtmltopdf.org (recommended) you can safely skip this section.

If you want to run `wkhtmltopdf` with an unpatched verison of webkit that requires
an X Window server, but your server (or Mac) does not have one installed,
you may find the `command_prefix` handy:

```Elixir
PdfGenerator.generate "<html..", command_prefix: "xvfb-run"
```

This can also be configured globally in your `config/config.exs`:

```Elixir
config :pdf_generator,
    command_prefix: "/usr/bin/xvfb-run"
```

If you will be generating multiple PDFs simultaneously, or in rapid succession,
you will need to configure `xvfb-run` to search for a free X server number,
or set the server number explicitly. You can use the `command_prefix` to pass
options to the `xvfb-run` command.

```Elixir
config :pdf_generator,
    command_prefix: ["xvfb-run", "-a"]
```

# Documentation

For more info, read the [docs on hex](http://hexdocs.pm/pdf_generator) or issue
`h PdfGenerator.generate` in your iex shell.

# Known issues

Unfortunately, with Elixir 1.7+ `System.cmd` seems to pass parameters
differently to the environment than it did before, now requiring shell options
like `--foo=bar` to be split up as `["--foo", "bar"]`. This behaviour seemingly
went away with OTP 22 in May 2019 and Elixir 1.8.2. So if you run into issues,
try upgrading to the latest Erlang/OTP and Elixir first, and do not hesitate
file a report.

# Contributing

Contributions (Issues, PRs…) are more than welcome. Please ave a quick read at 
the [Contribution tips](./CONTRIBUTING.md), though. It's basically about scope 
and kindness.


CHANGES.md
# Changes

  - 0.6.0
    - introducting `make` as build tool (optional) for chromium binaries
      (puppeteer)
    - **BUGFIX:** documentation: option `pagesize` requires string argument
      (for example `"letter"` or `"A4"`)
    - updated some npm dependencies for chromium
  - 0.5.7
    - **BUGFIX:** fix chrome-option parameter handling
  - 0.5.6
    - **BUGFIX:** fix A4 and A5 paper sizes in inches for **chrome-headless**:
      it's not 8.5 x 11.0 (US letter) but 8.26772 x 11.695 (DIN A4), the former
      being chrome-headless defaults. This is important if you want to create
      proper A4 pages
    - Users printing **US letter** sized PDFs, please use `page_size: "letter"`
  - 0.5.5
    - improved documentation on `prefer_system_executable: true` for chrome.
      Thanks to [Martin Richer](https://github.com/richeterre) for rasining this
      and a [PR](https://github.com/gutschilla/elixir-pdf-generator/pull/55)
    - improved documentation on `no_sandbox: true` for chrome in dockerized
      environment (running as root)
    - clarify that wkhtmltopdf installation snippet is for Ubuntu 18.04.
    - log call options as debug info to Logger
    - add "knows issues" section to README
  - 0.5.4
    - **BUGFIX** introduced in 0.5.0 that would crash `PdfGenerator.PathAgent`
      when chrome isn't found on path in certain situation. Thanks to
      [@radditude](https://github.com/radditude) for submitting a patch.
  - 0.5.3
    - **BUGFIX** introduced in 0.5.0 when certain shells don't accept
      `["foo=bar", …]` parameters which should correctly be `["foo", "bar"]`
      Thanks to [@egze](https://github.com/egze) for submitting a patch.
    - Refactored `PathAgent` that holds configuration state for readability and
      more fashionable and extensible error messages. Extensible towards new
      generators.
    - Updated README to be more elaborative on how to install `wkhtmltopdf` and
      `chrome-headless-render-pdf`
  - 0.5.2
    - **BUGFIX** introduced in 0.5.0 when global options to wkhtmltopdf weren't
      accepted any more due to wrong shell parameter order. Thanks to
      [manukall](https://github.com/manukall) for reporting.
  - 0.5.1
    - allow chrome to be executed as root via default config option
      `disable_chrome_sandbox` – this is required for an easy usage within a
      docker container as in
      [elixir-pdf-server](https://github.com/gutschilla/elixir-pdf-server)
  - 0.5.0
    - **Got rid of Porcelain** dependency as it interferes with many builds using
      plain `System.cmd/3`. Please note, that as of the documentation
      (https://hexdocs.pm/elixir/System.html#cmd/3) ports will be closed but in
      case wkhtmltopdf somehow hangs, nobody takes care of terminating it.
    - Refactored some sections
    - **Support URLs** instead of just plain HTML
    - **Support for chrome-headless** for (at least for me) faster and nicer renderings.
    - Since this is hopefully helpful, I rose the version to 0.5.0 even tough
      the API stays consistent
  - 0.4.0
    - Got rid of misc_random dependency. This was here to manage between
      depreciated random functions in Erlang. We go ahead using plain
      `Enum.random/1` instead, implementing our own
      `PdfGenerator.Random.string/1` function. This also removes a common
      pitfall when drafting a release with distillery.
      * Thanks to [Hugo Maia Vieira](https://github.com/hugomaiavieira) for this
        contribution!
      * Since `Enum.random/1` is only available since September 2015 (three
        years ago) I am OK with raising the minimum Elixir version to v1.1 –
        Since this may break projects still running on Elixir v1.0
        **I bumped the version to 0.4.0***.
  - 0.3.7
    - Adding in raise_on_missing_wkhtmltopdf_binary config, thanks
      to [veverkap](https://github.com/veverkap)
    - Document using xvfb-run with auto-servernum option, thanks
      to [Tony van Riet](https://github.com/tonyvanriet)
  - 0.3.6
    - bumped dependencies:
      * porcelain 2.0.3 to support newer erlangs and remove warnings
      * ex_doc 0.16 to remove warnings, remove from runtime
      * removed explixit earmark
    - add option to pick output pdf filename, thanks
      to [praveenperera](https://github.com/praveenperera)
    - improved README on heroku, corrected typos. Thanks
      to [jbhatab](https://github.com/jbhatab)
      and [maggy96](https://github.com/maggy96)
  - 0.3.5
    - add `generate_binary` and `generate_binary!` that immediately return the
      PDF binary instead of an `{:ok, filename}` tuple.
    - add `generate!` to immediately return the filename
    - some more tests
    - minor change `delete_temporary` must be truthy. (the old supported value
      `:html` will still work) and will delete both intermediate HTML And PDF
      files in `generate_binary` and `generate_binary!`
  - 0.3.5
    - add `generate_binary` and `generate_binary!` that immediately return the
      PDF binary instead of an `{:ok, filename}` tuple.
    - add `generate!` to immediately return the filename
    - some more tests
    - minor change `delete_temporary` must be truthy. (the old supported value
      `:html` will still work) and will delete both intermediate HTML And PDF
      files in `generate_binary` and `generate_binary!`
  - 0.3.4
    - BUGFIX: fix merge confusion to **really** support `xvfb-run` or other
      command prefixes to wkhtmltopdf
    - support explicit deletion of temporary files thanks to
      [Edipo Vinicius da Silva](https://github.com/edipox)
    - Improve README
  - 0.3.3
    - BUGFIX: typo in config/prod.exs
  - 0.3.2
    - support for command prefixes, most notabably **xvfb-run** to let a
      wkhtmltopdf which was compiled without an unpatched version of qt run on
      machines without an x server
    - (add in precompiled, patched binaries for wkhtmltopdf and libjpeg8 that are
      needed to run wkhtmltopdf without xvfb-run)
  - 0.3.1

config/test.exs
use Mix.Config

config :pdf_generator,
  command_prefix: "xvfb-run",
  raise_on_missing_wkhtmltopdf_binary: true


test/test_helper.exs
{:ok, _} = Application.ensure_all_started(:fake_server)
ExUnit.start()


test/pdf_generator_test.exs
defmodule PdfGeneratorTest do
  use ExUnit.Case

  @html "<html><body><h1>Hi</h1><p>Yikes!</p></body></html>"

  test "agent startup" do
    {:ok, _ } = File.stat PdfGenerator.PathAgent.get.wkhtml_path
  end

  test "basic PDF generation" do
    {:ok, temp_filename } = PdfGenerator.generate @html

    # File should exist and has some size
    file_info = File.stat! temp_filename
    assert file_info.size > 0
    pdf = File.read! temp_filename

    # PDF header should be present
    assert String.slice( pdf, 0, 6) == "%PDF-1"
  end

  test "command prefix with noop env" do
    {:ok, _temp_filename } = PdfGenerator.generate @html, command_prefix: "env"
  end

  test "command prefix with args with noop env" do
    {:ok, _temp_filename } = PdfGenerator.generate @html, command_prefix: ["env", "foo=bar"]
  end

  test "generate_binary reads file" do
    assert {:ok, "%PDF-1" <> _pdf} = @html |> PdfGenerator.generate_binary
  end

  test "chrome-headless from file" do
    {:ok, temp_filename } = PdfGenerator.generate(@html, generator: :chrome, page_size: "A5")

    # File should exist and has some size
    file_info = File.stat! temp_filename
    assert file_info.size > 0
    pdf = File.read! temp_filename

    # PDF header should be present
    assert String.slice( pdf, 0, 6) == "%PDF-1"
  end

  test "chrome-headless from URL (assuming google.com is up and running)" do
    {status, result} = PdfGenerator.generate({:url, "http://google.com"}, generator: :chrome)
    assert status == :ok
    assert result |> File.read! |> String.slice(0, 6) == "%PDF-1"
  end

  test "chrome's no-sandbox option doesn't crash" do
    {_status, result} = PdfGenerator.generate({:url, "http://google.com"}, generator: :chrome, disable_chrome_sandbox: true)
    assert result |> File.read! |> String.slice(0, 6) == "%PDF-1"
  end

  test "generate! returns a filename" do
    @html
    |> PdfGenerator.generate!
    |> File.exists?
    |> assert
  end

  test "generate! with filename option returns custom filename" do
    filename = PdfGenerator.generate!(@html, filename: "custom_file_name")
    assert File.exists?(filename)
    assert Path.basename(filename, ".pdf") == "custom_file_name"
  end

  test "generate_binary! reads file" do
    assert "%PDF-1" <> _pdf = @html |> PdfGenerator.generate_binary!
  end

  test "delete_temporary works" do
    # w/o delete_temporary, html should be there
    @html
    |> PdfGenerator.generate!
    |> String.replace( ~r(\.pdf$), ".html")
    |> File.exists?
    |> assert

    # with delete_temporary, html file should be gone
    @html
    |> PdfGenerator.generate!(delete_temporary: true)
    |> String.replace( ~r(\.pdf$), ".html")
    |> File.exists?()
    |> refute

    # cannot really be sure if temporyr file was deleted but this shouldn't
    # crash at least. We could scan the temp dir before and after but had to
    # make sure no other process wrote something in there which isn't exactly
    # robust.
    assert {:ok, "%PDF-1" <> _pdf} = @html |> PdfGenerator.generate_binary(delete_temporary: true)
  end

  test "shell_params are accepted" do
    @html
    |> PdfGenerator.generate!(shell_params: ["--orientation", "landscape", "--margin-right", "0"])
    |> File.exists?()
    |> assert
  end

end


mix.exs
defmodule PdfGenerator.Mixfile do
  use Mix.Project

  def project do
    [
      app: :pdf_generator,
      name: "PDF Generator",
      version: "0.6.0",
      elixir: ">= 1.1.0",
      deps: deps(),
      description: description(),
      package: package(),
      build_embedded: Mix.env == :prod,
      start_permanent: Mix.env == :prod
    ]
  end

  # Configuration for the OTP application
  #
  # Type `mix help compile.app` for more information
  def application do
    [
      applications: [:logger],
      mod: { PdfGenerator, [] }
    ]
  end

  def description do
    """
    A wrapper for wkhtmltopdf and chrome-headless(puppeteer) with optional
    support for encryption via pdftk.
    """
  end

  defp package do
    [
     files: ["lib", "mix.exs", "README.md", "LICENSE", "test"],
     maintainers: ["Martin Gutsch"],
     licenses: ["MIT"],
     links: %{
        "GitHub" => "https://github.com/gutschilla/elixir-pdf-generator"
      }
     ]
  end

  defp deps do
    [
        # communication with external programs
        {:ex_doc, "~> 0.19", only: :dev, runtime: false}
    ]
  end
end


lib/pdf_generator.ex
defmodule PdfGenerator do

  require Logger

  @vsn "0.6.0"

  @moduledoc """
  # PdfGenerator

  Provides a simple wrapper around [wkhtmltopdf](http://wkhtmltopdf.org) and
  [pdftk](https://www.pdflabs.com/tools/pdftk-the-pdf-toolkit/) to generate
  possibly encrypted PDFs from an HTML source.

  # Configuration (optional)

  if no or partial configuration is given, PdfGenerator will search for
  executables on path. This will rais an error when wkhtmltopdf cannot be
  found.

      config :pdf_generator,
            wkhtml_path: "/path/to/wkhtmltopdf",
            pdftk_path:  "/path/to/pdftk",


  In your config/config.exs. Add :pdf_generator to your mix.exs:
  Note that this is optional but advised to as it will perform a check on
  startup whether it can find a suitable wkhtmltopdf executable. It's
  generally better to have an app fail at startup than at later runtime.

    def application do
      [applications: [ .., :pdf_generator, ..], .. ]
    end

  If you don't want to autostart, issue

    PdfGenerator.start wkhtml_path: "/path/to/wkhtml_path"

  # System requirements

  - wkhtmltopdf or chrome-headless
  - pdftk (optional, for encrypted PDFs)

  Precompiled **wkhtmltopdf** binaries can be obtained here:
  http://wkhtmltopdf.org/downloads.html

  **pdftk** should be available as package on your system via

   - `apt-get install pdftk` on Debian/Ubuntu
   - `brew pdftk` on OSX (you'll need homebrew, of course)
   - Install the Exe-Installer on Windows found the project's homepage (link
   above)

  """

  use Application

  # See http://elixir-lang.org/docs/stable/elixir/Application.html
  # for more information on OTP Applications
  def start(_type, _args) do
      import Supervisor.Spec, warn: false

      children = [
        # Define workers and child supervisors to be supervised
        # worker(TestApp.Worker, [arg1, arg2, arg3])
        worker(
          PdfGenerator.PathAgent, [[
            wkhtml_path:                         Application.get_env(:pdf_generator, :wkhtml_path),
            pdftk_path:                          Application.get_env(:pdf_generator, :pdftk_path),
            raise_on_missing_wkhtmltopdf_binary: Application.get_env(:pdf_generator, :raise_on_missing_wkhtmltopdf_binary, true),
          ]]
        )
      ]

      opts = [strategy: :one_for_one, name: PdfGenerator.Supervisor]
      Supervisor.start_link(children, opts)
  end

  def defaults(), do: [generator: :wkhtmltopdf, page_size: "A4"]

  # return file name of generated pdf

  @doc """
  Generates a pdf file from given html string. Returns a string containing a
  temporary file path for that PDF.

  ## Options

   * `:generator` – either `chrome` or `wkhtmltopdf` (default)
   * `:prefer_system_executable` - set to `true` if you installed
     chrome-headless-render-pdf globally
   * `:no_sandbox` – disable sandbox for chrome, required to run as root (read: _docker_)
   * `:page_size` - output page size, defaults to "A4", other options are "letter" (US letter) and "A5"
   * `:open_password` - password required to open PDF. Will apply encryption to PDF
   * `:edit_password` - password required to edit PDF
   * `:shell_params` - list of command-line arguments to wkhtmltopdf or chrome
     see http://wkhtmltopdf.org/usage/wkhtmltopdf.txt for all options
   * `:delete_temporary` - true to remove the temporary html generated in
     the system tmp dir
   * `:filename` - filename you want for the output PDF (provide without .pdf extension),
     defaults to a random string

  # Examples

  pdf_path_1 = PdfGenerator.generate "<html><body><h1>Boom</h1></body></html>"
  pdf_path_2 = PdfGenerator.generate(
    "<html><body><h1>Boom</h1></body></html>",
    page_size:     "letter",
    open_password: "secret",
    edit_password: "g3h31m",
    shell_params: [ "--outline", "--outline-depth3", "3" ],
    delete_temporary: true,
    filename: "my_awesome_pdf"
  )
  """

  @type url           :: binary()
  @type html          :: binary()
  @type pdf_file_path :: binary()
  @type content       :: html | {:ok, url}
  @type reason        :: atom() | {atom(), any}
  @type opts          :: keyword()
  @type path          :: binary()
  @type html_path     :: path
  @type pdf_path      :: path
  @type generator     :: :wkhtmltopdf | :chrome

  @spec generate(content, opts) :: {:ok, pdf_file_path} | {:error, reason}
  def generate(content, opts \\ []) do

    options = Keyword.merge(defaults(), opts)

    generator     = options[:generator]

    open_password = options[:open_password]
    edit_password = options[:edit_password]
    delete_temp   = options[:delete_temporary]

    with {html_file, pdf_file}       <- make_file_paths(options),
         :ok                         <- maybe_write_html(content, html_file),
         {executable, arguments}     <- make_command(generator, options, content, {html_file, pdf_file}),
         {:cmd, {stderr, exit_code}} <- {:cmd, System.cmd(executable, arguments, stderr_to_stdout: true)},       # unfortuantely wkhtmltopdf returns 0 on errors as well :-/
         {:result_ok, true, _err}    <- {:result_ok, result_ok(generator, stderr, exit_code), stderr},           # so we inspect stderr instead
         {:rm, :ok}                  <- {:rm, maybe_delete_temp(delete_temp, html_file)},
         {:ok, encrypted_pdf}        <- maybe_encrypt_pdf(pdf_file, open_password, edit_password) do
      {:ok, encrypted_pdf}
    else
      {:error, reason}     -> {:error, reason}
      {:result_ok, _, err} -> {:error, {:generator_failed, err}}
      reason               -> {:error, reason}
    end
  end

  @spec maybe_write_html(content, path()) :: :ok | {:error, reason}
  def maybe_write_html({:url, _url}, _html_file_path),                      do: :ok
  def maybe_write_html({:html, html}, html_file_path),                      do: File.write(html_file_path, html)
  def maybe_write_html(html,          html_file_path) when is_binary(html), do: maybe_write_html({:html, html}, html_file_path)

  @spec make_file_paths(keyword()) :: {html_path, pdf_path}
  def make_file_paths(options) do
    filebase = options[:filename] |> generate_filebase()
    {filebase <> ".html", filebase <> ".pdf"}
  end

  def make_dimensions(options) when is_list(options) do
    options |> Enum.into(%{}) |> dimensions_for()
  end

  @doc ~s"""
  Returns `{width, height}` tuple for page sizes either as given or for A4 and
  A5. Defaults to A4 sizes. In inches. Because chrome wants imperial.
  """
  def dimensions_for(%{page_width: width, page_height: height}), do: {width, height}
  def dimensions_for(%{page_size: "A4"}),                        do: {"8.26772", "11.695"}
  def dimensions_for(%{page_size: "A5"}),                        do: {"5.8475",  "8.26772"}
  def dimensions_for(%{page_size: "letter"}),                    do: {"8.5",     "11"}
  def dimensions_for(_map),                                      do: dimensions_for(%{page_size: "A4"})

  @spec make_command(generator, opts, content, {html_path, pdf_path}) :: {path, list()}
  def make_command(:chrome, options, content, {html_path, pdf_path}) do
    chrome_executable  = PdfGenerator.PathAgent.get.chrome_path
    node_executable    = PdfGenerator.PathAgent.get.node_path
    disable_sandbox    = Application.get_env(:pdf_generator, :disable_chrome_sandbox) || options[:no_sandbox]

    # needs `make priv/node_modules` to be run when building
    priv_dir = :code.priv_dir(:pdf_generator) |> to_string()
    js_file  = "#{priv_dir}/node_modules/chrome-headless-render-pdf/dist/cli/chrome-headless-render-pdf.js"

    {executable, executable_args} =
      if options[:prefer_system_executable] && is_binary(chrome_executable) do
        {chrome_executable, []}
      else
        {node_executable, [js_file]}
      end

    {width, height} = make_dimensions(options)
    more_params = options[:shell_params] || []
    source =
      case content do
        {:url,  url} -> url
        _html        -> "file://" <> html_path
      end
    arguments = List.flatten([
      executable_args,
      [
        "--url", source,
        "--pdf", pdf_path,
        "--paper-width",   width,
        "--paper-height", height,
      ],
      more_params,
      if(disable_sandbox, do: ["--chrome-option=--no-sandbox"], else: [])
    ])
    {executable, arguments} |> inspect() |> Logger.debug()
    {executable, arguments}
  end

  def make_command(:wkhtmltopdf, options, content, {html_path, pdf_path}) do
    executable  = PdfGenerator.PathAgent.get.wkhtml_path
    source =
      case content do
        {:url, url} -> url
        _html       -> html_path
      end
    shell_params = options[:shell_params] || []
    arguments = List.flatten([
      shell_params,
      "--page-size", options[:page_size] || "A4",
      source, pdf_path
    ])
    # for wkhtmltopdf we support prefixes like ["xvfb-run", "-a"] to precede the actual command
    {executable, arguments} =
      case get_command_prefix(options) do
        nil                    -> {executable, arguments}
        [prefix | prefix_args] -> {prefix, prefix_args ++ [executable] ++ arguments}
        prefix                 -> {prefix, [executable | arguments]}
      end
    {executable, arguments} |> inspect() |> Logger.debug()
    {executable, arguments}
  end

  defp maybe_delete_temp(true,    file), do: File.rm(file)
  defp maybe_delete_temp(_falsy, _file), do: :ok

  def maybe_encrypt_pdf(pdf_file, open_password, edit_password)
  when is_binary(open_password) or is_binary(edit_password) do
    encrypt_pdf(pdf_file, open_password, edit_password)
  end

  def maybe_encrypt_pdf(pdf_file, _open_password, _edit_password) do
    {:ok, pdf_file}
  end

  defp result_ok(:chrome,     _string,          0), do: true
  defp result_ok(:chrome,     _string, _exit_code), do: false
  defp result_ok(:wkhtmltopdf, string, _exit_code), do: String.match?(string, ~r/Done/ms)

  defp get_command_prefix(options) do
    options[:command_prefix] || Application.get_env(:pdf_generator, :command_prefix)
  end

  defp generate_filebase(nil), do: generate_filebase(PdfGenerator.Random.string())
  defp generate_filebase(filename), do: Path.join(System.tmp_dir, filename)

  def encrypt_pdf(pdf_input_path, user_pw, owner_pw ) do
    pdftk_path      = PdfGenerator.PathAgent.get.pdftk_path
    pdf_output_file = Path.join System.tmp_dir, PdfGenerator.Random.string() <> ".pdf"

    pdftk_args = [
      pdf_input_path,
      "output", pdf_output_file,
      "owner_pw", random_if_undef(owner_pw),
      "user_pw",  random_if_undef(user_pw),
      "encrypt_128bit", "allow", "Printing", "CopyContents"
    ]

    {stderr, exit_code} = System.cmd(pdftk_path, pdftk_args, stderr_to_stdout: true)

    case exit_code do
      0 ->  {:ok, pdf_output_file}
      _ ->  {:error, {:pdftk, stderr}}
    end
  end

  defp random_if_undef(nil), do: PdfGenerator.Random.string(16)
  defp random_if_undef(any), do: any

  @doc """
  Takes same options as `generate` but will return an
  `{:ok, binary_pdf_content}` tuple.

  In case option _delete_temporary_ is true, will as well delete the temporary
  pdf file.
  """
  def generate_binary(html, options \\ []) do
    result = generate html, options
    case result do
      {:ok, filename}  -> {:ok, filename |> read_and_maybe_delete(options) }
      {:error, reason} -> {:error, reason}
    end
  end

  defp read_and_maybe_delete(filename, options) do
    content = filename |> File.read!
    if Keyword.get(options, :delete_temporary), do: filename |> File.rm
    content
  end

  @doc """
  Same as generate_binary but returns PDF content directly or raises on
  error.
  """
  def generate_binary!(html, options \\ []) do
    result = generate_binary html, options
    case result do
      {:ok, content}   -> content
      {:error, reason} -> raise "in-place generation failed: " <> reason
    end
  end

  @doc """
  Same as generate but returns PDF file name only (raises on error).
  """
  def generate!(html, options \\ []) do
    result = generate html, options
    case result do
      {:ok, filename}  -> filename
      {:error, reason} -> raise "HTML generation failed: " <> reason
    end
  end
end
